lib/espresso-runner.js

```diff
@@ -306,19 +354,26 @@ class EspressoRunner {
     try {
       await waitForCondition(async () => {
         if (hasSocketError) {
-          this.log.errorAndThrow(`Espresso server has failed to start due to an unexpected exception. ` +
+          throw this.log.errorWithException(
+            `Espresso server has failed to start due to an unexpected exception. ` +
             `Make sure the 'INTERNET' permission is requested in the Android manifest of your ` +
-            `application under test (<uses-permission android:name="android.permission.INTERNET" />)`);
+            `application under test (<uses-permission android:name="android.permission.INTERNET" />)`
+          );
         } else if (this.jwproxy.instrumentationState.exited) {
-          this.log.errorAndThrow(`Espresso server process has been unexpectedly terminated. ` +
-            `Check the Appium server log and the logcat output for more details`);
+          throw this.log.errorWithException(
+            `Espresso server process has been unexpectedly terminated. ` +
+            `Check the Appium server log and the logcat output for more details`
+          );
         }
+        let serverStatus;
         try {
-          await this.jwproxy.command('/status', 'GET');
-          return true;
+          (serverStatus = /** @type {import('@appium/types').StringRecord} */ (
+            await this.jwproxy.command('/status', 'GET')
+          ));
         } catch (e) {
           return false;
         }
+        return await this._verifyServerStatus(manifestPayload, serverStatus);
```

Comment:
this is actually the main target scenario for the check.

It covers cases where one tries to provide an app with precompiled server version incompatible with the current server version (we only throw an exception if major versions are different, for minor ones only a warning is shown).

It is also mandatory to verify the target package name to make sure we communicate with exactly the app we want to communicate and not something else which is by accident listening on the same port.

Reply:
I see, i agree with it basically.

I know of a person who uses `skipServerInstallation` to build an espresso server outside appium session since its appium running env has network limitation, which does not allow gradle deps download. He keeps using our vanilla espresso server, so it does not have its own customization.

I imagined that the espresso server itself had no breaking changes, but our JS side had a breaking change. Then, we would bump this driver's major version while the espresso server did not need to re-build with the latest code technically since no changes in the server side. Then, I wondered if we should require them to re-prepare another espresso server build with latest driver while they have the same espresso server code. It would make their server version management complicated a bit.

Then, maybe it would be nice to address this verification thing in `skipServerInstallation` caps description so that the running server must match with the required major version especially.

Reply:
makes sense. I have streamlined the logic, so in case the server or driver version cannot be parsed only a warning message would be printed. So one could simply set the version to an empty string or some non-parseable sentence in their local sources to prevent the driver from throwing an exception.

Regarding the documentation for the `skipServerInstallation` cap, would you prefer to add it in this PR or a separate one?

Reply:
Either is fine for me > would you...